tpm2/tpm2.go

```diff
@@ -0,0 +1,1247 @@
+}
+
+func decodeReadPublic(in []byte) ([]byte, []byte, []byte, error) {
+	var publicBlob []byte
+	var name []byte
+	var qualifiedName []byte
+
+	err := tpmutil.Unpack(in, &publicBlob, &name, &qualifiedName)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return publicBlob, name, qualifiedName, nil
+}
+
+// ReadPublic reads the public part of the object under handle.
+// Returns the public data, name and qualified name.
+func ReadPublic(rw io.ReadWriter, handle Handle) ([]byte, []byte, []byte, error) {
+	resp, err := runCommand(rw, tagNoSessions, cmdReadPublic, handle)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	publicBlob, name, qualifiedName, err := decodeReadPublic(resp)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return publicBlob, name, qualifiedName, nil
+}
+
+func encodeCreateKey(owner Handle, pcrNums []int, parentPassword, ownerPassword string, parms RSAParams) ([]byte, error) {
+	b1, err := tpmutil.Pack(Handle(owner))
+	if err != nil {
+		return nil, err
+	}
+	b2, err := tpmutil.Pack([]byte(nil))
+	if err != nil {
+		return nil, err
+	}
+	b3, err := encodePasswordAuthArea(parentPassword, HandlePasswordSession)
+	if err != nil {
+		return nil, err
+	}
+	t1, err := encodePasswordData(ownerPassword)
+	if err != nil {
+		return nil, err
+	}
+	b4, err := encodeSensitiveArea(t1[2:], []byte(nil))
+	if err != nil {
+		return nil, err
+	}
+	b5, err := encodeRSAParams(parms)
+	if err != nil {
+		return nil, err
+	}
+	b6, err := tpmutil.Pack([]byte(nil))
+	if err != nil {
+		return nil, err
+	}
+	b7, err := encodeLongPCR(1, pcrNums)
+	if err != nil {
+		return nil, err
+	}
+	args := append(b1, b2...)
+	args = append(args, b3...)
+	args = append(args, b4...)
+	args = append(args, b5...)
+	args = append(args, b6...)
+	args = append(args, b7...)
+	return args, nil
+}
+
+func decodeCreateKey(in []byte) ([]byte, []byte, error) {
+	var tpm2bPrivate []byte
+	var tpm2bPublic []byte
+
+	err := tpmutil.Unpack(in[4:], &tpm2bPrivate, &tpm2bPublic)
+	if err != nil {
+		return nil, nil, err
+	}
+	return tpm2bPrivate, tpm2bPublic, nil
+}
+
+// CreateKey creates a new RSA key pair under the owner handle.
+// Returns private key and public key blobs.
+func CreateKey(rw io.ReadWriter, owner Handle, pcrNums []int, parentPassword, ownerPassword string, parms RSAParams) ([]byte, []byte, error) {
+	cmd, err := encodeCreateKey(owner, pcrNums, parentPassword, ownerPassword, parms)
+	if err != nil {
+		return nil, nil, err
+	}
+	resp, err := runCommand(rw, tagSessions, cmdCreate, tpmutil.RawBytes(cmd))
+	if err != nil {
+		return nil, nil, err
+	}
+	privateBlob, publicBlob, err := decodeCreateKey(resp)
+	if err != nil {
+		return nil, nil, err
+	}
+	return privateBlob, publicBlob, nil
+}
+
+func encodeLoad(parentHandle Handle, parentAuth, ownerAuth string, publicBlob, privateBlob []byte) ([]byte, error) {
+	b1, err := tpmutil.Pack(parentHandle)
+	if err != nil {
+		return nil, err
+	}
+	b3, err := encodePasswordData(parentAuth)
```

Comment:
What happened to `b2`? Is this a copy-paste issue? These two calls (`encodePasswordData`, `encodePasswordAuthArea`) seem like good candidates for refactoring as well.